core/variant_parser.cpp

```diff
@@ -1754,7 +1754,7 @@ Error VariantWriter::write(const Variant &p_variant, StoreStringFunc p_store_str
 					}
 
 					p_store_string_func(p_store_string_ud, "\"" + E->get().name + "\":");
-					write(obj->get(E->get().name), p_store_string_func, p_store_string_ud, p_encode_res_func, p_encode_res_ud);
+					write(obj->get(E->get().name), p_store_string_func, p_store_string_ud, p_encode_res_func, p_encode_res_ud, p_recursion_count);
```

Comment:
Looks like objects also need the recursion check.

<details>
<summary>3.x</summary>

```gdscript
class Test extends Resource:
    export var a: Resource

func _run() -> void:
    var test := Test.new()
    test.a = test
    print(var2str(test))
```

</details>

<details>
<summary>4.x</summary>

```gdscript
class Test extends Resource:
    @export var a: Resource

func _run() -> void:
    var test := Test.new()
    test.a = test
    print(var_to_str(test))
```

</details>